Blog post debezium iceberg (#714)
debezium/debezium.github.io: _posts/2021-10-01-debezium-server-iceberg.adoc

---
layout: post
title:  Using Debezium to Create ACID Data Lake
date:   2021-10-01
tags: [ debezium, iceberg, datalake, lakehouse ]
author: isimsek
---

> Do you need to build Lakehouse with near realtime data pipeline, do you want it to support ACID transactions and updates on data lake?

Reviewer: I think we need set the context a bit here, not all the Debezium users are familiar with the concepts of data lakes and such. Heck, I'm not even sure myself what "lake house" means in this context :)

> It is possible with "Debezium Server Iceberg" without any dependency to kafka or spark applications.

Reviewer: ```suggestion
It is possible with "Debezium Server Iceberg" without any dependency to Apache Kafka or Apache Spark applications.
```

==== Apache Iceberg

https://iceberg.apache.org/[Apache Iceberg] is an open table format for huge analytic datasets.
Iceberg adds tables to compute engines including Spark, Trino, PrestoDB, Flink and Hive using a high-performance table format that works just like a SQL table.
It supports ACID Inserts, Row level Deletes/Updates. It has flexible foundation and comes with https://iceberg.apache.org[many features].

== Debezium Server Iceberg

**Debezium Server Iceberg** project implements new consumer which uses Iceberg Java API to consume events.
Iceberg consumer converts received events to iceberg format and commits them to destination iceberg table. With the consumer it's possible to use any cloud store/any catalog supported by iceberg.

on high level data processing works like below ::
for each batch of events
* it groups events per event destination,
* for each destination, events are converted to iceberg records. at this step event schema used to map data to iceberg record (`debezium.format.value.schemas.enable` should be enabled(true)).

> * After events converted to iceberg records, they are saved to iceberg parquet files(iceberg data and delete files(for upsert)).

Reviewer: There should be a sentence about what Parquet is; the double nested braces are a bit confusing, can this be reworded?

> * as a last step these files are committed to destination table(uploaded to destination storage)) using iceberg java API.

Reviewer: ```suggestion
* as a last step, these files are committed to the destination tables (i.e. uploaded to the destination storage) using  the Iceberg java API.
```

> If destination table not found in the destination catalog consumer will try to create it using event schema and key schema(record key/PK).

Reviewer: ```suggestion
If the destination table is not found in the destination catalog, the consumer will try to create it using the event schema and key schema(record key/PK).
```

What is a catalog? This needs explanation.

(There are quite a few "the"s missing in general; perhaps you can do a general review pass and add them? If needed, I also can help with that).

image::../assets/images/debezium-iceberg.png[Debezium-Iceberg]

> Currently, Iceberg Consumer supports only json events. With json events event schema must be enabled to do correct type conversion to iceberg record.

Reviewer: Ok, this explains my earlier question. I think this should be pulled up in order to achieve a comprehensible flow.

> example configuration::

Reviewer: "Here is a complete example configuration for using Debezium Server with the Iceberg adaptor:"

[source,properties]
----
debezium.sink.type=iceberg
# run with append mode
debezium.sink.iceberg.upsert=false
debezium.sink.iceberg.upsert-keep-deletes=true
# iceberg
debezium.sink.iceberg.table-prefix=debeziumcdc_
debezium.sink.iceberg.table-namespace=debeziumevents
debezium.sink.iceberg.fs.defaultFS=s3a://S3_BUCKET);
debezium.sink.iceberg.warehouse=s3a://S3_BUCKET/iceberg_warehouse
debezium.sink.iceberg.type=hadoop
debezium.sink.iceberg.catalog-name=mycatalog
debezium.sink.iceberg.catalog-impl=org.apache.iceberg.hadoop.HadoopCatalog
# enable event schemas
debezium.format.value.schemas.enable=true
debezium.format.value=json
# complex nested data types are not supported, do event flattening. unwrap message!
debezium.transforms=unwrap
debezium.transforms.unwrap.type=io.debezium.transforms.ExtractNewRecordState
debezium.transforms.unwrap.add.fields=op,table,source.ts_ms,db
debezium.transforms.unwrap.delete.handling.mode=rewrite
debezium.transforms.unwrap.drop.tombstones=true
----

> === Update, Append modes

Reviewer: ```suggestion
=== Update and Append Modes
```

Would that be better?

> By default, Iceberg sink is running with upsert mode `debezium.sink.iceberg.upsert=true`. When a row updated on source table destination row replaced with the new updated version, and deleted records are deleted from destination. With upsert mode data at destination kept identical to source data. Update mode uses iceberg equality delete feature and creates delete files using key of the debezium event(PK). To avoid duplicate data deduplication is done on each batch and only last version of the record kept.

Reviewer: Some words seem to be missing in the second sentence and the one after that. What is a delete file? And how would data duplication happen? I don't fully grok that section.

> Note: For the tables without record key(PK) operation mode falls back to `append` even configuration is set to upsert mode.

Reviewer: Instead of this pathological case, can you perhaps describe how to use append mode in general? I think that's interesting for many users who would like to keep the entire change history for a record in Iceberg.

==== Keeping Deleted Records

For some use cases it's useful to keep deleted records as soft deletes, this is possible by setting `debezium.sink.iceberg.upsert-keep-deletes` to `true`.
This setting will keep the latest version of deleted record (`__op=d`) in the iceberg table. Setting it to false will remove deleted records from the destination table.

==== Append mode

This is most straightforward operation mode, setting `debezium.sink.iceberg.upsert` to `false` sets the operation mode to append.

> With append mode data deduplication is not done and all received records are appended to destination table.

Reviewer: Ok, I think I understand now what you mean with deduplication here. But there's a difference between actual deduplication (i.e. the same change event is emitted twice, e.g. in case after a connector crash, and this exact duplicate is discarded, as e.g. recognizable via the event's LSN) and keeping only the latest state for reach record (vs. keeping all the events for that record). I think this deserves some clarification.

=== Optimizing batch size (commit interval)

Debezium extracts/consumes database events in real time and this could cause too frequent commits( generate too many small files) to iceberg table,
which is not optimal for batch processing especially when near realtime data feed is sufficient.
To avoid this problem it's possible to use following configuration and increase batch size per commit.

> **MaxBatchSizeWait**: uses debezium metrics to optimize batch size, it periodically reads streaming queue current size and waits until it reaches to `max.batch.size`.

Reviewer: ```suggestion
**MaxBatchSizeWait**: uses Debezium metrics to optimize batch size, it periodically reads streaming queue current size and waits until it reaches to `max.batch.size`.
```
(here and elsewhere)

Maximum wait and check intervals are controlled by `debezium.sink.batch.batch-size-wait.max-wait-ms`, `debezium.sink.batch.batch-size-wait.wait-interval-ms` properties.
During the wait debezium events are collected in memory (in debezium streaming queue), this way each commit receives more records and consistent batch size.
Note: this setting should be configured together with `debezium.source.max.queue.size` and `debezium.source.max.batch.size` debezium properties

Note: It's also possible to do data compaction using iceberg, compacting data and metadata files to get the best performance.

> example setting::

Reviewer: Perhaps only show those settings applying to compaction?

[source,properties]
----
debezium.sink.batch.batch-size-wait=MaxBatchSizeWait
debezium.sink.batch.metrics.snapshot-mbean=debezium.postgres:type=connector-metrics,context=snapshot,server=testc
debezium.sink.batch.metrics.streaming-mbean=debezium.postgres:type=connector-metrics,context=streaming,server=testc
debezium.source.connector.class=io.debezium.connector.postgresql.PostgresConnector
debezium.source.max.batch.size=50000
debezium.source.max.queue.size=400000
debezium.sink.batch.batch-size-wait.max-wait-ms=60000
debezium.sink.batch.batch-size-wait.wait-interval-ms=10000
----

> === Destination, iceberg catalog

Reviewer: Having a section title for a single sentence seems a bit too fine-grained. It is a general question though which is a bit unclear to me: how are the files actually stored in Iceberg? Using some object storage like S3? Can I store it in my local filesystem too? Can we have an example for this somehow?

Iceberg consumer uses iceberg catalog to read and commit data to destination table, destination could be any cloud storage and any catalog supported by iceberg.

> == Next Datawarehouse, Curated layer

Reviewer: The title sounds a bit odd, what exactly are you trying to express here?

Now we got perfect raw layer with near realtime data feed, we can build Curated Layer,Analytic Layer or Datawarehouse on top of.

For example someone could easily use https://iceberg.apache.org/spark-writes/[Spark SQL](or Prestodb, Trino, Flink) and process this data to next layer

> loading slowly changing dimension type 2:

Reviewer: What is dimension type 2?
